File: services/auth-service/src/utils/jwt.util.ts
import jwt from 'jsonwebtoken';
import { TokenError } from '@jam/base-app';

/**
 * JWT Utility
 *
 * Handles JWT token generation and validation
 */

export interface TokenPayload {
  userId: string;
  email: string;
  type: 'access' | 'refresh' | 'reset';
}

export interface TokenPair {
  token: string;
  refreshToken: string;
  expiresIn: number;
  expiresAt: Date;
}

export class JWTUtil {
  private static readonly SECRET = process.env.JWT_SECRET || 'your-secret-key-change-in-production';
  private static readonly REFRESH_SECRET = process.env.JWT_REFRESH_SECRET || 'your-refresh-secret-key';
  private static readonly ACCESS_TOKEN_EXPIRY = '15m'; // 15 minutes
  private static readonly REFRESH_TOKEN_EXPIRY = '7d'; // 7 days

  /**
   * Generate access and refresh token pair
   */
  static generateTokenPair(userId: string, email: string): TokenPair {
    const accessToken = this.generateAccessToken(userId, email);
    const refreshToken = this.generateRefreshToken(userId, email);

    // Calculate expiry time (15 minutes)
    const expiresIn = 15 * 60; // seconds
    const expiresAt = new Date(Date.now() + expiresIn * 1000);

    return {
      token: accessToken,
      refreshToken,
      expiresIn,
      expiresAt,
    };
  }

  /**
   * Generate access token
   */
  static generateAccessToken(userId: string, email: string): string {
    const payload: TokenPayload = {
      userId,
      email,
      type: 'access',
    };

    return jwt.sign(payload, this.SECRET, {
      expiresIn: this.ACCESS_TOKEN_EXPIRY,
      issuer: 'jam-auth-service',
      audience: 'jam-app',
    });
  }

  /**
   * Generate refresh token
   */
  static generateRefreshToken(userId: string, email: string): string {
    const payload: TokenPayload = {
      userId,
      email,
      type: 'refresh',
    };

    return jwt.sign(payload, this.REFRESH_SECRET, {
      expiresIn: this.REFRESH_TOKEN_EXPIRY,
      issuer: 'jam-auth-service',
      audience: 'jam-app',
    });
  }

  /**
   * Verify access token
   */
  static verifyAccessToken(token: string): TokenPayload {
    try {
      const payload = jwt.verify(token, this.SECRET, {
        issuer: 'jam-auth-service',
        audience: 'jam-app',
      }) as TokenPayload;

      if (payload.type !== 'access') {
        throw new TokenError('Invalid token type');
      }

      return payload;
    } catch (error: any) {
      if (error.name === 'TokenExpiredError') {
        throw new TokenError('Token has expired');
      } else if (error.name === 'JsonWebTokenError') {
        throw new TokenError('Invalid token');
      } else {
        throw new TokenError(error.message);
      }
    }
  }

  /**
   * Verify refresh token
   */
  static verifyRefreshToken(token: string): TokenPayload {
    try {
      const payload = jwt.verify(token, this.REFRESH_SECRET, {
        issuer: 'jam-auth-service',
        audience: 'jam-app',
      }) as TokenPayload;

      if (payload.type !== 'refresh') {
        throw new TokenError('Invalid token type');
      }

      return payload;
    } catch (error: any) {
      if (error.name === 'TokenExpiredError') {
        throw new TokenError('Refresh token has expired');
      } else if (error.name === 'JsonWebTokenError') {
        throw new TokenError('Invalid refresh token');
      } else {
        throw new TokenError(error.message);
      }
    }
  }

  /**
   * Generate reset token (short-lived for password resets)
   */
  static generateResetToken(userId: string, email: string): string {
    const payload: TokenPayload = {
      userId,
      email,
      type: 'reset',
    };

    return jwt.sign(payload, this.SECRET, {
      expiresIn: '15m', // 15 minutes
      issuer: 'jam-auth-service',
      audience: 'jam-app',
    });
  }

  /**
   * Verify reset token
   */
  static verifyResetToken(token: string): TokenPayload {
    try {
      const payload = jwt.verify(token, this.SECRET, {
        issuer: 'jam-auth-service',
        audience: 'jam-app',
      }) as TokenPayload;

      if (payload.type !== 'reset') {
        throw new TokenError('Invalid token type');
      }

      return payload;
    } catch (error: any) {
      if (error.name === 'TokenExpiredError') {
        throw new TokenError('Reset token has expired');
      } else if (error.name === 'JsonWebTokenError') {
        throw new TokenError('Invalid reset token');
      } else {
        throw new TokenError(error.message);
      }
    }
  }

  /**
   * Decode token without verification (for debugging)
   */
  static decode(token: string): any {
    return jwt.decode(token);
  }

  /**
   * Get token expiry time
   */
  static getTokenExpiry(token: string): Date | null {
    const decoded = jwt.decode(token) as any;
    if (!decoded || !decoded.exp) {
      return null;
    }

    return new Date(decoded.exp * 1000);
  }

  /**
   * Check if token is expired
   */
  static isExpired(token: string): boolean {
    const expiry = this.getTokenExpiry(token);
    if (!expiry) {
      return true;
    }

    return expiry < new Date();
  }
}

File: services/todos-service/backend/src/utils/jwt_util.py
from datetime import datetime, timedelta
from typing import Dict, Optional
from jose import JWTError, jwt
from ..config import settings


class JWTUtil:
    """JWT token generation and validation"""

    @staticmethod
    def create_access_token(user_id: str, email: str) -> str:
        """Create access token"""
        expires_delta = timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
        expire = datetime.utcnow() + expires_delta

        to_encode = {
            "sub": user_id,
            "email": email,
            "type": "access",
            "iss": "jam-auth-service",
            "aud": "jam-app",
            "exp": expire,
            "iat": datetime.utcnow(),
        }

        encoded_jwt = jwt.encode(
            to_encode,
            settings.JWT_SECRET,
            algorithm=settings.JWT_ALGORITHM
        )
        return encoded_jwt

    @staticmethod
    def create_refresh_token(user_id: str, email: str) -> str:
        """Create refresh token"""
        expires_delta = timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
        expire = datetime.utcnow() + expires_delta

        to_encode = {
            "sub": user_id,
            "email": email,
            "type": "refresh",
            "iss": "jam-auth-service",
            "aud": "jam-app",
            "exp": expire,
            "iat": datetime.utcnow(),
        }

        encoded_jwt = jwt.encode(
            to_encode,
            settings.JWT_REFRESH_SECRET,
            algorithm=settings.JWT_ALGORITHM
        )
        return encoded_jwt

    @staticmethod
    def create_reset_token(user_id: str, email: str) -> str:
        """Create password reset token"""
        expires_delta = timedelta(minutes=15)
        expire = datetime.utcnow() + expires_delta

        to_encode = {
            "sub": user_id,
            "email": email,
            "type": "reset",
            "iss": "jam-auth-service",
            "aud": "jam-app",
            "exp": expire,
            "iat": datetime.utcnow(),
        }

        encoded_jwt = jwt.encode(
            to_encode,
            settings.JWT_SECRET,
            algorithm=settings.JWT_ALGORITHM
        )
        return encoded_jwt

    @staticmethod
    def verify_token(token: str, token_type: str = "access") -> Optional[Dict]:
        """Verify and decode token"""
        try:
            secret = settings.JWT_REFRESH_SECRET if token_type == "refresh" else settings.JWT_SECRET

            payload = jwt.decode(
                token,
                secret,
                algorithms=[settings.JWT_ALGORITHM],
                issuer="jam-auth-service",
                audience="jam-app"
            )

            if payload.get("type") != token_type:
                return None

            return payload
        except JWTError:
            return None

    @staticmethod
    def get_token_expiry(token: str) -> Optional[datetime]:
        """Get token expiration time"""
        try:
            payload = jwt.decode(
                token,
                settings.JWT_SECRET,
                algorithms=[settings.JWT_ALGORITHM],
                options={"verify_signature": False}
            )
            exp = payload.get("exp")
            if exp:
                return datetime.fromtimestamp(exp)
            return None
        except JWTError:
            return None
